libfwupdplugin/fu-fmap-firmware.c

```diff
@@ -0,0 +1,278 @@
+		return FALSE;
+	}
+
+	if (i + fmap_size ((struct fmap *)&image[i]) > len)
+		return FALSE;
+
+	*offset = i;
+	return TRUE;
+}
+
+static gint
+popcnt (guint u)
+{
+	gint count;
+
+	/* K&R method */
+	for (count = 0; u; count++)
+		u &= (u - 1);
+
+	return count;
+}
+
+static gboolean
+fmap_find (const guint8 *image, gsize image_len, gsize *offset, GError **error)
+{
+	gboolean ret;
+
+	if ((image == NULL) || (image_len == 0)) {
+		g_set_error_literal (error,
+				     G_IO_ERROR,
+				     G_IO_ERROR_INVALID_DATA,
+				     "invalid image");
+		return FALSE;
+	}
+
+	if (popcnt (image_len) == 1)
+		ret = fmap_bsearch (image, image_len, offset, error);
+	else
+		ret = fmap_lsearch (image, image_len, offset, error);
+
+	if (!ret) {
+		g_prefix_error (error, "failed fmap_find: ");
+		return FALSE;
+	}
+
+	return TRUE;
+}
+
+static gboolean
+fu_fmap_firmware_parse (FuFirmware *firmware,
+			GBytes *fw,
+			guint64 addr_start,
+			guint64 addr_end,
+			FwupdInstallFlags flags,
+			GError **error)
+{
+	gsize image_len;
+	guint8 *image = (guint8 *)g_bytes_get_data (fw, &image_len);
+	gsize offset;
+	const struct fmap *fmap;
+	gsize i;
+
+	/* corrupt */
+	if (g_bytes_get_size (fw) < sizeof (struct fmap)) {
+		g_set_error_literal (error,
+				     G_IO_ERROR,
+				     G_IO_ERROR_INVALID_DATA,
+				     "firmware too small for fmap");
+		return FALSE;
+	}
+
+	if (!fmap_find (image, image_len, &offset, error)) {
+		g_set_error_literal (error,
+				     G_IO_ERROR,
+				     G_IO_ERROR_INVALID_DATA,
+				     "cannot find fmap in image");
+		return FALSE;
+	}
+
+	fmap = (const struct fmap *)(image + offset);
+
+	if (fmap->size != image_len) {
+		g_set_error (error,
+			     G_IO_ERROR,
+			     G_IO_ERROR_INVALID_DATA,
+			     "file size incorrect, expected 0x%04x got 0x%04x",
+			     (guint) fmap->size,
+			     (guint) image_len);
+		return FALSE;
+
+	}
+
+	if (fmap->nareas < 1) {
+		g_set_error (error,
+			     G_IO_ERROR,
+			     G_IO_ERROR_INVALID_DATA,
+			     "number of areas too small, got %" G_GUINT16_FORMAT,
+			     fmap->nareas);
+		return FALSE;
+	}
+
+	for (i = 0; i < fmap->nareas; i++) {
+		const struct fmap_area *area = &fmap->areas[i];
+		g_autoptr(FuFirmwareImage) img = NULL;
+		g_autoptr(GBytes) bytes = NULL;
+
+		img = fu_firmware_image_new (NULL);
+		bytes = g_bytes_new_from_bytes (fw,
+						(gsize) area->offset,
+						(gsize) area->size);
+		fu_firmware_image_set_id (img, (const gchar *) area->name);
+		fu_firmware_image_set_idx (img, i + 1);
+		fu_firmware_image_set_addr (img, (guint64) area->offset);
+		fu_firmware_image_set_bytes (img, bytes);
+		fu_firmware_add_image (firmware, img);
+	}
+
+	/* success */
+	return TRUE;
+}
+
+static void
+fu_fmap_firmware_init (FuFmapFirmware *self)
+{
+}
+
+static void
+fu_fmap_firmware_class_init (FuFmapFirmwareClass *klass)
+{
+	FuFirmwareClass *klass_firmware = FU_FIRMWARE_CLASS (klass);
+	klass_firmware->parse = fu_fmap_firmware_parse;
+}
+
+FuFirmware *
```

Comment:
This will need gtk doc documentation to indicate which version of fwupd it's landing in (1.5.0)

Reply:
Thanks! I was wondering why the CIs were upset.